spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/subscriber/PubSubSubscriberTemplate.java

```diff
@@ -214,19 +314,63 @@ public Subscriber subscribe(String subscription,
 
 		List<AcknowledgeablePubsubMessage> ackableMessages = pull(pullRequest);
 
-		if (ackableMessages.size() > 0) {
+		if (!ackableMessages.isEmpty()) {
 			ack(ackableMessages);
 		}
 
 		return ackableMessages.stream().map(AcknowledgeablePubsubMessage::getPubsubMessage)
 				.collect(Collectors.toList());
 	}
 
+	@Override
+	public ListenableFuture<List<PubsubMessage>> pullAndAckAsync(String subscription, Integer maxMessages,
+			Boolean returnImmediately) {
+		Assert.hasText(subscription, "The subscription can't be null or empty.");
+
+		if (maxMessages != null) {
+			Assert.isTrue(maxMessages > 0, "The maxMessages must be greater than 0.");
+		}
+
+		PullRequest pullRequest = this.subscriberFactory.createPullRequest(
+				subscription, maxMessages, returnImmediately);
+
+		final SettableListenableFuture<List<PubsubMessage>> settableFuture = new SettableListenableFuture<>();
+
+		this.pullAsync(pullRequest).addCallback(
```

Comment:
All of the code above this line in the method is shared with `pullAndAck`. Can we try to reduce the code duplication?

Reply:
I moved the validation into the `SubscriberFactory.createPullRequest`, where the `subscription` validation also already was done. This removed the duplication.